### PR TITLE
Bug 1377620 - Elasticsearch bulk indexer ignores changes that happened before newest additions

### PR DESCRIPTION
**Pull request by dylanwh**

### CODE REVIEW

File: Bugzilla/Elastic/Indexer.pm [hunk 1]
```diff
@@ -109,7 +104,7 @@ sub _bulk_helper {
     );
 }
 
-sub find_largest_mtime {
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: Was this a cosmetic change uncovered while fixing the core problem, but not necessarily related to the problem itself?

File: Bugzilla/Elastic/Indexer.pm [hunk 2]
```diff
@@ -170,54 +165,44 @@ sub _debug_sql {
...
-
-    my $updated_ids = $self->_select_updated_ids($class);
-    if ($updated_ids) {
-        $self->_bulk_load_ids($bulk, $class, $updated_ids) if @$updated_ids;
-    }
+    my $bulk        = $self->_bulk_helper($class);
+    my $last_mtime  = $self->_find_largest_mtime($class);
+    my $last_id     = $self->_find_largest_id($class);
+    my $new_ids     = $self->_select_all_ids($class, $last_id);
+    my $updated_ids = $self->_select_updated_ids($class, $last_mtime);
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: This is so much easier to read. Thank you. Now I understand what's going on.

File: Bugzilla/Elastic/Indexer.pm [hunk 3]
```diff
@@ -170,54 +165,44 @@ sub _debug_sql {
...
-    my $dbh   = Bugzilla->dbh;
-    my $mtime = $self->find_largest_mtime($class);
-    if ($mtime && $mtime != $self->mtime) {
-        my ($updated_sql, $updated_params) = $self->_debug_sql($class->ES_SELECT_UPDATED_SQL($mtime));
-        return $dbh->selectcol_arrayref($updated_sql, undef, @$updated_params);
-    } else {
-        return undef;
-    }
+    my $dbh = Bugzilla->dbh;
+    my ($updated_sql, $updated_params) = $self->_debug_sql($class->ES_SELECT_UPDATED_SQL($last_mtime));
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: What validates `$last_mtime` as being a valid mtime at all?

File: Bugzilla/Elastic/Indexer.pm [hunk 4]
```diff
@@ -170,54 +165,44 @@ sub _debug_sql {
...
     $self->put_mapping($class);
-    $self->clear_mtime;
     $self->_bulk_load_ids($self->_bulk_helper($class), $class, $ids);
 }
 
 sub _bulk_load_ids {
     my ($self, $bulk, $class, $all_ids) = @_;
 
     my $iter  = natatime $class->ES_OBJECTS_AT_ONCE, @$all_ids;
-    my $mtime = $self->mtime;
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: Why is the lazy-cache here removed? Is that related to the bug somehow?

Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: it was sort of wrong.
You could forget to clear it and then have the mtime for users get inserted into the mixed into the mtimes for comments... the new code is much easier to follow.

File: Bugzilla/Elastic/Indexer.pm (hunk 1 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: It was because a previous review asked for sanity checking a value, and a desire to not repeat myself. So I merged two code paths into one implementation

File: Bugzilla/Elastic/Indexer.pm (hunk 3 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: it's checked to be a number, which is good enough for mysql, elasticsearch, and perl

File: Bugzilla/Elastic/Indexer.pm [hunk 5]
```diff
@@ -104,34 +104,41 @@ sub _bulk_helper {
...
 
-    my $result = $self->client->search(
-        index => $self->index_name,
-        type  => $class->ES_TYPE,
-        body  => {
-            aggs => { $class->ID_FIELD => { extended_stats => { field => $class->ID_FIELD } } },
-            size => 0
-        }
-    );
+    return $self->_find_largest('es_mtime');
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: Is a `$class->ES_MTIME_FIELD` available here?

Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: no, because mtime doesn't come from the object, but from the indexer itself.
